File: apps/api/app/schemas/sns.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ─── SNS Settings ────────────────────────────────────────────

class SnsSettingOut(BaseModel):
    id: UUID
    store_id: UUID

    twitter_enabled: bool
    twitter_api_key: Optional[str] = None
    twitter_api_secret: Optional[str] = None
    twitter_access_token: Optional[str] = None
    twitter_access_secret: Optional[str] = None

    instagram_enabled: bool
    instagram_account_id: Optional[str] = None
    instagram_access_token: Optional[str] = None

    line_enabled: bool
    line_channel_token: Optional[str] = None
    line_channel_secret: Optional[str] = None

    auto_new_arrival: bool
    auto_price_down: bool
    auto_sold_out: bool

    new_arrival_template: str
    price_down_template: str
    sold_out_template: str

    repost_enabled: bool
    repost_interval_weeks: int
    repost_platforms: Optional[List[str]] = None

    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SnsSettingUpdate(BaseModel):
    twitter_enabled: Optional[bool] = None
    twitter_api_key: Optional[str] = None
    twitter_api_secret: Optional[str] = None
    twitter_access_token: Optional[str] = None
    twitter_access_secret: Optional[str] = None

    instagram_enabled: Optional[bool] = None
    instagram_account_id: Optional[str] = None
    instagram_access_token: Optional[str] = None

    line_enabled: Optional[bool] = None
    line_channel_token: Optional[str] = None
    line_channel_secret: Optional[str] = None

    auto_new_arrival: Optional[bool] = None
    auto_price_down: Optional[bool] = None
    auto_sold_out: Optional[bool] = None

    new_arrival_template: Optional[str] = None
    price_down_template: Optional[str] = None
    sold_out_template: Optional[str] = None

    repost_enabled: Optional[bool] = None
    repost_interval_weeks: Optional[int] = Field(None, ge=1, le=12)
    repost_platforms: Optional[List[str]] = None


# ─── SNS Posts ───────────────────────────────────────────────

class SnsPostOut(BaseModel):
    id: UUID
    store_id: UUID
    car_id: Optional[UUID] = None

    trigger: str
    platform: str
    status: str

    caption: str
    image_urls: Optional[List[str]] = None
    posted_at: Optional[datetime] = None
    error_message: Optional[str] = None
    repost_count: int

    created_at: datetime

    class Config:
        from_attributes = True


class SnsPostCreate(BaseModel):
    car_id: Optional[UUID] = None
    trigger: str = "manual"
    platform: str = "all"
    caption: str
    image_urls: Optional[List[str]] = None


class SnsPostListOut(BaseModel):
    items: List[SnsPostOut]
    total: int


# ─── Preview ─────────────────────────────────────────────────

class SnsPreviewOut(BaseModel):
    trigger: str
    caption: str


# ─── Repost Schedule ─────────────────────────────────────────

class RepostScheduleItem(BaseModel):
    car_id: UUID
    car_name: str
    last_posted_at: datetime
    next_repost_at: datetime
    overdue: bool

File: apps/web/src/lib/api/sns.ts
import { apiFetch } from "@/lib/api";

// ─── 型定義 ──────────────────────────────────────────────────

export type SnsSetting = {
  id: string;
  store_id: string;

  twitter_enabled: boolean;
  twitter_api_key: string | null;
  twitter_api_secret: string | null;
  twitter_access_token: string | null;
  twitter_access_secret: string | null;

  instagram_enabled: boolean;
  instagram_account_id: string | null;
  instagram_access_token: string | null;

  line_enabled: boolean;
  line_channel_token: string | null;
  line_channel_secret: string | null;

  auto_new_arrival: boolean;
  auto_price_down: boolean;
  auto_sold_out: boolean;

  new_arrival_template: string;
  price_down_template: string;
  sold_out_template: string;

  repost_enabled: boolean;
  repost_interval_weeks: number;
  repost_platforms: string[] | null;

  created_at: string;
  updated_at: string;
};

export type SnsSettingUpdate = Partial<Omit<SnsSetting, "id" | "store_id" | "created_at" | "updated_at">>;

export type SnsPost = {
  id: string;
  store_id: string;
  car_id: string | null;
  trigger: string;
  platform: string;
  status: "pending" | "posted" | "failed" | "skipped";
  caption: string;
  image_urls: string[] | null;
  posted_at: string | null;
  error_message: string | null;
  repost_count: number;
  created_at: string;
};

export type SnsPostListOut = {
  items: SnsPost[];
  total: number;
};

export type SnsPostCreate = {
  car_id?: string | null;
  trigger?: string;
  platform?: string;
  caption: string;
  image_urls?: string[] | null;
};

export type SnsPreview = {
  trigger: string;
  caption: string;
};

export type RepostScheduleItem = {
  car_id: string;
  car_name: string;
  last_posted_at: string;
  next_repost_at: string;
  overdue: boolean;
};

// ─── API 関数 ────────────────────────────────────────────────

export async function getSnsSettings(): Promise<SnsSetting> {
  return apiFetch<SnsSetting>("/api/v1/sns/settings");
}

export async function updateSnsSettings(body: SnsSettingUpdate): Promise<SnsSetting> {
  return apiFetch<SnsSetting>("/api/v1/sns/settings", { method: "PUT", body });
}

export type SnsPostFilter = {
  trigger?: string;
  platform?: string;
  status?: string;
  limit?: number;
  offset?: number;
};

export async function listSnsPosts(params?: SnsPostFilter): Promise<SnsPostListOut> {
  const qs = new URLSearchParams();
  if (params?.trigger) qs.set("trigger", params.trigger);
  if (params?.platform) qs.set("platform", params.platform);
  if (params?.status) qs.set("status", params.status);
  if (params?.limit != null) qs.set("limit", String(params.limit));
  if (params?.offset != null) qs.set("offset", String(params.offset));
  const q = qs.toString();
  return apiFetch<SnsPostListOut>(`/api/v1/sns/posts${q ? "?" + q : ""}`);
}

export async function createSnsPost(body: SnsPostCreate): Promise<SnsPost> {
  return apiFetch<SnsPost>("/api/v1/sns/posts", { method: "POST", body });
}

export async function retrySnsPost(postId: string): Promise<SnsPost> {
  return apiFetch<SnsPost>(`/api/v1/sns/posts/${postId}/retry`, { method: "POST" });
}

export async function previewSnsPost(carId: string): Promise<SnsPreview[]> {
  return apiFetch<SnsPreview[]>(`/api/v1/sns/preview?car_id=${carId}`);
}

export async function getRepostSchedule(): Promise<RepostScheduleItem[]> {
  return apiFetch<RepostScheduleItem[]>("/api/v1/sns/repost-schedule");
}

export async function triggerSnsPost(body: SnsPostCreate & { trigger: string }): Promise<SnsPost> {
  return apiFetch<SnsPost>("/api/v1/sns/trigger", { method: "POST", body });
}
